c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  // promise returned from `get_return_object()` is destroyed while this coroutine is suspended.
+};
+
+template <typename Self, typename T>
+class CoroutineBase {
+public:
+  void return_value(T value) {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill(kj::mv(value));
+  }
+};
+template <typename Self>
+class CoroutineBase<Self, void> {
+public:
+  void return_void() {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill();
+  }
+};
+// The Coroutines spec has no `_::FixVoid<T>` equivalent to unify valueful and valueless co_return
+// statements, and programs are ill-formed if the coroutine implementation object (Coroutine<T>) has
+// both a `return_value()` and `return_void()`. No amount of EnableIffery can get around it, so
+// these return_* functions live in a CRTP mixin.
+
+template <typename T>
+class Coroutine<T>::Adapter: private Event {
+public:
+  explicit Adapter(PromiseFulfiller<T>& f, Coroutine::Handle coroutine)
+      : fulfiller(f), coroutine(coroutine) {
+    coroutine.promise().adapter = this;
+  }
+  KJ_DISALLOW_COPY(Adapter);
+
+  ~Adapter() noexcept(false) {
+    // If a coroutine runs to completion, its frame will already have been destroyed by the time the
+    // adapter is destroyed, so this if-condition will almost always be false.
+    if (coroutine) {
+      // coroutine.promise() is a reference to our implementation object, Coroutine<T>.
+      coroutine.promise().adapter = nullptr;
+      coroutine.destroy();
+    }
+  }
+
+  void coroutineDestroyed() {
+    // The coroutine was destroyed before the adapter.
+    coroutine = nullptr;
+  }
+
+  void awaitSuspend(PromiseNode& nodeParam, ExceptionOrValue& resultParam) {
+    // Called by Awaiter::await_suspend() to tell us to wait for `node` to become ready, then store
+    // the node's result in `result`.
+
+    // We'll need to save these references for fire(). Note that Awaiter is movable, and `result` is
+    // its subobject, but it is safe to store a reference to `result` in Adapter, because by the
+    // time `await_suspend()` is called, Awaiter has already been returned from `await_transform()`
+    // and stored in its final resting place in the coroutine's frame.
+    node = &nodeParam;
+    result = &resultParam;
+
+    // We use the promise adapter as an Event, because we must be able to destroy the coroutine from
+    // within the fire() callback. The promise adapter can do this, because it lives outside of the
+    // coroutine's frame, unlike Coroutine<T> or Awaiter<U>.
+    node->onReady(this);
+  }
+
+  PromiseFulfiller<T>& fulfiller;
+  // Used to fulfill/reject the enclosing coroutine's kj::Promise<T>.
+
+private:
+  Maybe<Own<Event>> fire() override {
+    // The promise this coroutine is currently waiting on is ready. Extract its result and check for
+    // exceptions.
+
+    // Destroying or resuming the coroutine will leave these pointers dangling, so we call this
```

Comment:
I guess this is just a safety measure? The pointers won't actually be used for anything, right?

I am conflicted. On one hand it's just two stores, on the other hand any code here is going to be specialized many times driving a lot of code bloat.

Reply:
Yeah, it's purely a safety measure. I think it'd be reasonable to remove it, or hide it behind KJ_DEBUG.

Reply:
Removed: https://github.com/capnproto/capnproto/pull/1104/commits/8b8850f96cb9a43bb7e098578abd3a528975aca8
And then completely obviated when I gave up trying to support -fno-exceptions: https://github.com/capnproto/capnproto/pull/1104/commits/23721723d988ea9a523e10e3543b6f5a3d7f0f18